File: backend/app/harness/agents/query_plan_agent.py
"""
QueryPlan Agent — Agent-First 查询规划。

两种模式：
- plan()          : 单次 LLM 调用，快但不自校验（legacy, fallback）
- agentic_plan() : Tool-using loop，用 search_preview 试查自校验（推荐）

LLM 不可用时 returns None → 调用方回退到 build_query() fallback。
"""
import json
import logging
import re
from typing import Dict, List, Optional, Set

from app.harness.prompts.query_plan import build_query_plan_prompt
from app.services.prompt_loader import load_prompt

logger = logging.getLogger(__name__)


# ──────────────────────── Tools for Agentic Loop ────────────────────────

async def _tool_search_preview(query: str, source: str = "local_kb") -> dict:
    """
    Agentic tool: 试查一个查询，返回命中数 + 前 5 个标题。
    Agent 用这个观察自己的查询效果，决定是否调整或 finalize。
    """
    if not query or not query.strip():
        return {"error": "empty query"}
    try:
        from app.services.fetchers.international import ALL_FETCHERS
        fetcher = ALL_FETCHERS.get(source)
        if not fetcher:
            return {"error": f"source '{source}' unavailable"}
        results = await fetcher.fetch(query=query.strip(), max_results=5)
        return {
            "count": len(results),
            "top_titles": [
                (r.get("title") or "")[:100] for r in results[:5]
            ],
        }
    except Exception as e:
        return {"error": str(e)[:200]}


# ──────────────────────── Agentic Loop System Prompt ────────────────────────
# System prompt + 参数（max_iterations / preview_source）都在
# backend/app/prompts/agents/query_plan_agentic.md 里维护。
# 改 md 后无需重启 worker —— mtime 热重载。


class QueryPlanAgent:
    """Agent-First query planning, supports both single-shot and tool-using modes."""

    def __init__(self, llm_manager=None):
        self._llm = llm_manager

    # ═══════════════════════════════════════════════════════════════════
    # Agentic Mode (tool-using, self-verifying)
    # ═══════════════════════════════════════════════════════════════════

    async def agentic_plan(
        self,
        project_description: str,
        memory_text: str = "",
        max_iterations: Optional[int] = None,
    ):
        """
        Tool-using agent loop:
        LLM 用 search_preview 自己试查、观察命中、调整、finalize。

        模型变强时 agent 自动变强，不依赖硬规则。

        System prompt + max_iterations 从 prompts/agents/query_plan_agentic.md 读取（mtime 热重载）。

        Returns:
            QueryPlan with optional _rationale / _clarification_needed / _clarification_message
            None if LLM unavailable or loop exceeds budget
        """
        if not self._llm:
            logger.warning("[AgenticPlan] LLM 不可用")
            return None

        prompt_file = load_prompt("agents/query_plan_agentic")
        if max_iterations is None:
            max_iterations = int(prompt_file.get("max_iterations", 5) or 5)

        history = [
            {
                "role": "user",
                "content": (
                    f"【研究描述】\n{project_description}\n\n"
                    + (f"【历史记忆】\n{memory_text}\n\n" if memory_text else "")
                    + "请开始规划。先 search_preview 试你判断的核心概念。"
                ),
            }
        ]

        for iteration in range(max_iterations):
            prompt = self._render_conversation(history, system_prompt=prompt_file.body)
            response = await self._llm.generate(
                prompt, temperature=0.1,
                response_format={"type": "json_object"},
            )

            if not response:
                logger.warning("[AgenticPlan] iter=%d LLM 返回空", iteration)
                return None

            action = _parse_action_json(response)
            if not action:
                logger.warning(
                    "[AgenticPlan] iter=%d 解析失败: %s", iteration, response[:200]
                )
                history.append({"role": "assistant", "content": response[:500]})
                history.append({
                    "role": "user",
                    "content": "你的回复不是合法 JSON。请严格输出一个 JSON 对象，无任何其它文字。",
                })
                continue

            history.append({
                "role": "assistant",
                "content": json.dumps(action, ensure_ascii=False),
            })

            act_name = action.get("action")

            if act_name == "search_preview":
                query = action.get("query", "")
                source = action.get("source", "local_kb")
                result = await _tool_search_preview(query, source)
                logger.info(
                    "[AgenticPlan] iter=%d preview(%r) → count=%s",
                    iteration, query[:50], result.get("count"),
                )
                history.append({
                    "role": "user",
                    "content": f"[tool_result]\n{json.dumps(result, ensure_ascii=False)}",
                })
                continue

            if act_name == "finalize":
                plan_data = action.get("plan") or {}
                logger.info(
                    "[AgenticPlan] iter=%d FINALIZE: base_query=%r clarification=%s",
                    iteration,
                    (plan_data.get("base_query") or "")[:60],
                    plan_data.get("clarification_needed"),
                )
                return self._build_query_plan(plan_data)

            # Unknown action
            history.append({
                "role": "user",
                "content": "未知 action。只能用 search_preview 或 finalize。",
            })

        logger.warning("[AgenticPlan] 超出预算 %d 次仍未 finalize", max_iterations)
        return None

    def _render_conversation(self, history: List[Dict], system_prompt: str) -> str:
        """把对话历史渲染成一个大 prompt（用于不支持 native messages 的 provider）。"""
        parts = [system_prompt, ""]
        for msg in history:
            role = msg["role"].upper()
            parts.append(f"=== {role} ===")
            parts.append(msg["content"])
            parts.append("")
        parts.append("=== ASSISTANT ===")
        return "\n".join(parts)

    def _build_query_plan(self, plan_data: dict):
        """把 agent finalize 的 dict 转成 QueryPlan dataclass。"""
        from app.services.query_builder import QueryPlan
        from datetime import datetime

        base_query = (plan_data.get("base_query") or "").strip()

        # clarification path: build minimal plan to surface the message
        clarification_needed = bool(plan_data.get("clarification_needed"))
        clarification_message = plan_data.get("clarification_message") or ""

        if not base_query and not clarification_needed:
            return None

        current_year = datetime.now().year
        year_to = plan_data.get("year_to")
        if not isinstance(year_to, int) or year_to > current_year + 1:
            year_to = current_year
        year_from = plan_data.get("year_from")
        if year_from is not None and (not isinstance(year_from, int) or year_from < 1900):
            year_from = None

        language_scope = plan_data.get("language_scope", "international")
        if language_scope not in ("chinese_first", "international", "global"):
            language_scope = "international"

        qp = QueryPlan(
            base_query=base_query or "placeholder",
            expanded_terms=[w for w in base_query.split() if len(w) >= 2],
            exclude_terms=plan_data.get("exclude_terms", []) or [],
            year_from=year_from,
            year_to=year_to,
            sources=["local_kb"],  # search_mode filter will override later
            max_results_per_source=plan_data.get("max_per_source", 20),
            language_scope=language_scope,
            original_chinese_query=plan_data.get("chinese_query"),
        )
        qp._rationale = (plan_data.get("rationale") or "")[:300]
        qp._clarification_needed = clarification_needed
        qp._clarification_message = clarification_message[:500]
        return qp

    # ═══════════════════════════════════════════════════════════════════
    # Legacy Single-shot Mode (fallback)
    # ═══════════════════════════════════════════════════════════════════

    async def plan(
        self,
        project_description: str,
        memory_text: str = "",
        round_number: int = 1,
        max_rounds: int = 5,
        tool_reliability: Dict = None,
        prev_source_stats: Dict = None,
        disabled_sources: Set[str] = None,
    ):
        """
        Legacy single-shot plan. Kept as fallback when agentic_plan fails.
        """
        if not self._llm:
            logger.warning("[QueryPlanAgent] LLM 不可用，需要 fallback")
            return None

        if disabled_sources is None:
            from app.services.source_config_store import get_effective_disabled
            disabled_sources = await get_effective_disabled()

        system_prompt, user_prompt = build_query_plan_prompt(
            project_description=project_description,
            memory_text=memory_text,
            round_number=round_number,
            max_rounds=max_rounds,
            tool_reliability=tool_reliability or {},
            disabled_sources=disabled_sources,
            prev_source_stats=prev_source_stats,
        )

        combined = f"{system_prompt}\n\n---\n\n{user_prompt}"

        for attempt in range(2):
            try:
                result = await self._llm.generate(
                    combined, temperature=0.2,
                    response_format={"type": "json_object"},
                )
                if not result:
                    if attempt == 0:
                        continue
                    return None

                parsed = _parse_query_plan(result, disabled_sources)
                if parsed:
                    from app.services.query_builder import QueryPlan

                    qp = QueryPlan(
                        base_query=parsed["base_query"],
                        expanded_terms=parsed["expanded_terms"],
                        exclude_terms=parsed["exclude_terms"],
                        year_from=parsed["year_from"],
                        year_to=parsed["year_to"],
                        sources=parsed["sources"],
                        max_results_per_source=parsed["max_per_source"],
                        language_scope=parsed["language_scope"],
                        original_chinese_query=parsed.get("chinese_query"),
                    )
                    qp._rationale = parsed.get("rationale", "")

                    logger.info(
                        "[QueryPlanAgent.plan] 规划成功: query='%s', %d sources, %s",
                        qp.base_query[:50],
                        len(qp.sources),
                        f"{qp.year_from}-{qp.year_to}",
                    )
                    return qp

                if attempt == 0:
                    logger.warning("[QueryPlanAgent.plan] 解析失败，重试: %s", result[:150])
                    continue

            except Exception as e:
                logger.warning("[QueryPlanAgent.plan] 规划异常 (attempt %d): %s", attempt + 1, e)
                if attempt == 0:
                    continue

        return None


# ──────────────────────── JSON Parsing Helpers ────────────────────────

def _parse_action_json(text: str) -> Optional[dict]:
    """Extract a valid JSON 'action' object from agent response.

    Tolerates:
    - Markdown code fences (```json ... ```)
    - Leading/trailing explanations
    - Nested braces (uses balanced brace scanning, not greedy regex)
    - Multiple candidate objects (picks the first one with 'action' key)
    """
    if not text:
        return None
    text = text.strip()
    # Strip markdown code fences
    if text.startswith("```"):
        text = re.sub(r"^```(?:json)?\s*\n?", "", text)
        text = re.sub(r"\n?\s*```\s*$", "", text)

    # Scan for balanced JSON objects
    candidates = []
    depth = 0
    start = -1
    for i, ch in enumerate(text):
        if ch == "{":
            if depth == 0:
                start = i
            depth += 1
        elif ch == "}":
            depth -= 1
            if depth == 0 and start >= 0:
                candidates.append(text[start : i + 1])
                start = -1

    for candidate in candidates:
        try:
            obj = json.loads(candidate)
        except json.JSONDecodeError:
            continue
        if isinstance(obj, dict) and "action" in obj:
            return obj

    # Fallback: last resort greedy match (might parse partial)
    match = re.search(r"\{[\s\S]*\}", text)
    if match:
        try:
            return json.loads(match.group())
        except json.JSONDecodeError:
            return None
    return None


def _parse_query_plan(text: str, disabled_sources: Set[str]) -> Optional[Dict]:
    """解析 legacy plan() 的 LLM JSON 输出。"""
    match = re.search(r'\{[\s\S]*"base_query"[\s\S]*\}', text)
    if not match:
        match = re.search(r'\{[\s\S]+\}', text)
        if not match:
            return None

    try:
        data = json.loads(match.group())
    except json.JSONDecodeError:
        return None

    base_query = data.get("base_query")
    if not base_query or not isinstance(base_query, str) or len(base_query) < 3:
        return None

    sources = data.get("sources", [])
    if not isinstance(sources, list) or len(sources) < 1:
        sources = ["openalex", "crossref", "arxiv"]
    sources = [s for s in sources if s not in disabled_sources]
    if not sources:
        sources = ["openalex"]

    expanded = data.get("expanded_terms", [])
    if not isinstance(expanded, list):
        expanded = []
    base_words = [w for w in base_query.split() if len(w) >= 2]
    for w in base_words:
        if w.lower() not in [e.lower() for e in expanded]:
            expanded.insert(0, w)

    from datetime import datetime
    current_year = datetime.now().year

    year_to = data.get("year_to", current_year)
    if not isinstance(year_to, int) or year_to > current_year + 1:
        year_to = current_year

    year_from = data.get("year_from")
    if year_from is not None:
        if not isinstance(year_from, int) or year_from < 1900:
            year_from = None

    max_per_source = data.get("max_per_source", 20)
    if not isinstance(max_per_source, int) or max_per_source < 5:
        max_per_source = 20

    language_scope = data.get("language_scope", "international")
    if language_scope not in ("chinese_first", "international", "global"):
        language_scope = "international"

    return {
        "base_query": base_query.strip(),
        "chinese_query": data.get("chinese_query"),
        "expanded_terms": expanded,
        "exclude_terms": data.get("exclude_terms", []) or [],
        "year_from": year_from,
        "year_to": year_to,
        "sources": sources,
        "max_per_source": max_per_source,
        "language_scope": language_scope,
        "rationale": str(data.get("rationale", ""))[:200],
    }

File: client/src/data/agents/queryPlanAgent.ts
/**
 * QueryPlan Agent — 客户端版（移植自 backend `app/harness/agents/query_plan_agent.py`）。
 *
 * 两种模式：
 * - `legacyPlan()`: 单次 LLM 调用，快但不自校验（fallback 用）
 * - `agenticPlan()`: Tool-using loop，用 `searchPreview` 试查自校验（推荐）
 *
 * 三层兜底（任务规范 / `feedback_llm_parser_fallback.md`）：
 * 1. agentic LLM JSON 解析失败 → 重试一次（在同一 loop 里追加纠正提示）
 * 2. agentic 仍失败（耗尽 maxIterations 或始终解析不出）→ 自动 fallback 调 `legacyPlan`
 * 3. `legacyPlan` 也失败 → throw `QueryPlanError`
 *
 * 与 backend 差异：
 * - backend `QueryPlan` dataclass 把 sources 拍成一个列表共享 base_query；客户端 spec 要求
 *   每个 source 独立 `{keywords, filters}`，所以这里把 base_query 切词后均摊到所有 sources，
 *   filters 装 year_from/year_to/language_scope/exclude_terms。
 * - 不读 `disabled_sources` Redis 缓存，调用方传进来的 `sources` 列表即真相。
 * - `searchPreview` 工具调用走 `fetcherApi`（**TODO**：客户端 fetcher API 还未实现，目前所有
 *   测试通过 `vi.mock` 注入；正式上线时需要在 `client/src/api/fetcher.ts` 里补一个
 *   `searchPreview(source, keywords)` 端点，或直接在客户端跑 OpenAlex/arXiv 公开 API）。
 */
import type { LLMResult } from '../llm/types'
import { loadPrompt } from './promptLoader'

// ──────────────────────── Types ────────────────────────

/** 单一 source 的检索方案（每个 source 一份）。 */
export interface SourcePlan {
  keywords: string[]
  filters: Record<string, unknown>
}

/** 完整 QueryPlan 输出。 */
export interface QueryPlan {
  /** 每个 source 一份 plan（key=source name）。 */
  perSource: Record<string, SourcePlan>
  /** Agent 给出的 rationale / clarification 信息汇总。 */
  reasoning: string
  /** Agentic loop 实际跑了几次（legacy 路径恒为 0）。 */
  iterations: number
  /** 透传 agent finalize 时的额外字段（base_query、language_scope 等），方便 caller 二次加工。 */
  meta: {
    baseQuery: string
    chineseQuery?: string | null
    yearFrom?: number | null
    yearTo?: number | null
    languageScope: 'chinese_first' | 'international' | 'global'
    excludeTerms: string[]
    clarificationNeeded: boolean
    clarificationMessage: string
    /** 走了哪条路径，便于排查。 */
    mode: 'agentic' | 'legacy'
  }
}

/** Agentic loop 的中间 action JSON 形态。 */
type AgentAction =
  | {
      action: 'search_preview'
      query: string
      source?: string
    }
  | {
      action: 'finalize'
      plan: {
        base_query?: string
        chinese_query?: string | null
        year_from?: number | null
        year_to?: number | null
        language_scope?: string
        rationale?: string
        clarification_needed?: boolean
        clarification_message?: string
        exclude_terms?: string[]
      }
    }
  | { action: string; [k: string]: unknown }

/** Legacy 单次 plan 的 LLM JSON 形态。 */
interface LegacyPlanJson {
  base_query?: string
  chinese_query?: string | null
  expanded_terms?: string[]
  exclude_terms?: string[]
  year_from?: number | null
  year_to?: number | null
  sources?: string[]
  max_per_source?: number
  language_scope?: string
  rationale?: string
}

// ──────────────────────── LLM Manager 接口（duck-typed） ────────────────────────

/**
 * 仅依赖 `generate(prompt, options)` 一个方法，方便测试时注入 mock。
 * 与 `client/src/data/llm/manager.ts` 的命名空间 export 兼容。
 */
export interface LLMLike {
  generate(
    prompt: string,
    options?: {
      temperature?: number
      response_format?: { type: 'json_object' | 'text' } | null
    },
  ): Promise<LLMResult | null>
}

// ──────────────────────── fetcherApi 占位 ────────────────────────

/**
 * 客户端 fetcher API 占位类型。
 *
 * **TODO（B5+）**：客户端目前没有 `searchPreview` 实现，正式启用 agentic 模式之前需要：
 * - 选项 A：在 sp-api 加 `POST /api/search/preview` 端点
 * - 选项 B：客户端直接 fetch OpenAlex/arXiv/Crossref 公开 API（绕开 sp-api）
 *
 * 单测中通过 `vi.mock` 注入实现。
 */
export interface FetcherApiLike {
  searchPreview(
    source: string,
    keywords: string,
  ): Promise<{ count: number; topTitles: string[] } | { error: string }>
}

/** 默认 fetcherApi 实现（throw not implemented，强制调用方 inject）。 */
const _DEFAULT_FETCHER_API: FetcherApiLike = {
  async searchPreview() {
    return { error: 'fetcherApi.searchPreview not implemented (B5 TODO)' }
  },
}

// ──────────────────────── Errors ────────────────────────

export class QueryPlanError extends Error {
  constructor(
    message: string,
    public readonly cause?: unknown,
  ) {
    super(message)
    this.name = 'QueryPlanError'
  }
}

// ──────────────────────── JSON Parsing Helpers ────────────────────────

/**
 * 从 agent 回复里提取一个含有 `action` 字段的 JSON 对象。
 *
 * 兼容：
 * - markdown ```json fences
 * - 前后解释文字
 * - 嵌套花括号（用平衡扫描，不用贪婪正则）
 * - 多个候选对象（取第一个含 `action` 的）
 *
 * 对齐 backend `_parse_action_json`。
 */
export function parseActionJson(text: string | null | undefined): AgentAction | null {
  if (!text) return null
  let body = text.trim()
  if (body.startsWith('```')) {
    body = body.replace(/^```(?:json)?\s*\n?/, '')
    body = body.replace(/\n?\s*```\s*$/, '')
  }

  // 平衡扫描所有顶层 {...} 对象
  const candidates: string[] = []
  let depth = 0
  let start = -1
  for (let i = 0; i < body.length; i++) {
    const ch = body[i]
    if (ch === '{') {
      if (depth === 0) start = i
      depth++
    } else if (ch === '}') {
      depth--
      if (depth === 0 && start >= 0) {
        candidates.push(body.slice(start, i + 1))
        start = -1
      }
    }
  }

  for (const cand of candidates) {
    try {
      const obj = JSON.parse(cand)
      if (obj && typeof obj === 'object' && 'action' in obj) {
        return obj as AgentAction
      }
    } catch {
      continue
    }
  }

  // Last-resort: greedy 取第一个 {...} 区段（可能截断）
  const m = body.match(/\{[\s\S]*\}/)
  if (m) {
    try {
      return JSON.parse(m[0]) as AgentAction
    } catch {
      return null
    }
  }
  return null
}

/** 解析 legacy plan 的 LLM JSON。对齐 backend `_parse_query_plan`。 */
export function parseLegacyPlanJson(text: string | null | undefined): LegacyPlanJson | null {
  if (!text) return null
  // 优先匹配含 base_query 字段的对象，避免把别的 JSON 错认成 plan
  let match = text.match(/\{[\s\S]*"base_query"[\s\S]*\}/)
  if (!match) match = text.match(/\{[\s\S]+\}/)
  if (!match) return null

  let data: LegacyPlanJson
  try {
    data = JSON.parse(match[0]) as LegacyPlanJson
  } catch {
    return null
  }
  if (!data.base_query || typeof data.base_query !== 'string' || data.base_query.length < 3) {
    return null
  }
  return data
}

// ──────────────────────── QueryPlanAgent ────────────────────────

export interface AgenticPlanParams {
  projectDescription: string
  /** 历史记忆原文（可选）。 */
  memorySnapshot?: string
  /** 想要规划的 sources 列表（key 必须能在 fetcherApi 里查得到）。 */
  sources: string[]
  /** Agentic loop 上限，默认从 prompt frontmatter 读，否则 5。 */
  maxIterations?: number
}

export interface LegacyPlanParams {
  projectDescription: string
  memorySnapshot?: string
  sources: string[]
  /** 透传给 legacy prompt 的轮次信息（默认 1/5）。 */
  roundNumber?: number
  maxRounds?: number
}

export class QueryPlanAgent {
  constructor(
    private readonly llm: LLMLike,
    private readonly fetcherApi: FetcherApiLike = _DEFAULT_FETCHER_API,
  ) {}

  // ═══════════════════════════════════════════════════════════════════
  // Agentic Mode
  // ═══════════════════════════════════════════════════════════════════

  /**
   * Agentic loop 规划。
   *
   * 三层兜底：
   * 1. JSON parse 失败 → 在 history 里追加纠正提示，下一轮 LLM 重试
   * 2. 耗尽 `maxIterations` 仍未 finalize → fallback 调 `legacyPlan`
   * 3. `legacyPlan` 也失败 → throw `QueryPlanError`
   */
  async agenticPlan(params: AgenticPlanParams): Promise<QueryPlan> {
    const { projectDescription, memorySnapshot = '', sources } = params
    if (!projectDescription || !projectDescription.trim()) {
      throw new QueryPlanError('projectDescription is required')
    }
    if (!Array.isArray(sources) || sources.length === 0) {
      throw new QueryPlanError('sources must be a non-empty array')
    }

    const promptFile = loadPrompt('agents/query_plan_agentic')
    const maxIterations
      = params.maxIterations
      ?? (promptFile.get<number>('max_iterations', 5) as number)
      ?? 5
    const previewSource = promptFile.get<string>('preview_source', 'local_kb') as string

    type Message = { role: 'user' | 'assistant'; content: string }
    const history: Message[] = [
      {
        role: 'user',
        content:
          `【研究描述】\n${projectDescription}\n\n`
          + (memorySnapshot ? `【历史记忆】\n${memorySnapshot}\n\n` : '')
          + '请开始规划。先 search_preview 试你判断的核心概念。',
      },
    ]

    let iterations = 0
    let lastParseError: string | null = null

    for (let i = 0; i < maxIterations; i++) {
      iterations = i + 1
      const prompt = this.renderConversation(history, promptFile.body)

      let response: LLMResult | null
      try {
        response = await this.llm.generate(prompt, {
          temperature: 0.1,
          response_format: { type: 'json_object' },
        })
      } catch (e) {
        return this.fallbackToLegacy(params, `LLM threw: ${(e as Error).message}`)
      }

      if (!response || !response.text) {
        return this.fallbackToLegacy(params, 'LLM returned empty response')
      }

      const action = parseActionJson(response.text)
      if (!action) {
        // 三层兜底 1：parse 失败，让 LLM 在下一轮纠正
        lastParseError = response.text.slice(0, 200)
        history.push({ role: 'assistant', content: response.text.slice(0, 500) })
        history.push({
          role: 'user',
          content: '你的回复不是合法 JSON。请严格输出一个 JSON 对象，无任何其它文字。',
        })
        continue
      }

      // 成功 parse,清掉上一次的 parse error 标记
      lastParseError = null
      history.push({ role: 'assistant', content: JSON.stringify(action) })

      const actName = (action as { action: string }).action

      if (actName === 'search_preview') {
        const a = action as { query?: string; source?: string }
        const query = (a.query ?? '').trim()
        const source = a.source ?? previewSource
        if (!query) {
          history.push({ role: 'user', content: '[tool_result]\n{"error":"empty query"}' })
          continue
        }
        let result: { count: number; topTitles: string[] } | { error: string }
        try {
          result = await this.fetcherApi.searchPreview(source, query)
        } catch (e) {
          result = { error: (e as Error).message?.slice(0, 200) ?? 'unknown' }
        }
        history.push({
          role: 'user',
          content: `[tool_result]\n${JSON.stringify(result)}`,
        })
        continue
      }

      if (actName === 'finalize') {
        const a = action as Extract<AgentAction, { action: 'finalize' }>
        const plan = this.buildQueryPlan(a.plan ?? {}, sources, 'agentic', iterations)
        if (plan) return plan
        // finalize 解析出来 base_query 都没有 → fallback
        return this.fallbackToLegacy(params, 'finalize plan missing base_query')
      }

      // 未知 action：提醒并继续
      history.push({
        role: 'user',
        content: '未知 action。只能用 search_preview 或 finalize。',
      })
    }

    // 三层兜底 2：耗尽 iterations 仍未 finalize → fallback
    return this.fallbackToLegacy(
      params,
      lastParseError
        ? `agentic loop exhausted (last parse error: ${lastParseError})`
        : `agentic loop exhausted after ${maxIterations} iterations`,
    )
  }

  // ═══════════════════════════════════════════════════════════════════
  // Legacy Mode
  // ═══════════════════════════════════════════════════════════════════

  /**
   * 单次 LLM plan（无 search preview 自检）。
   *
   * 失败时 retry 一次；两次都失败 throw `QueryPlanError`。
   */
  async legacyPlan(params: LegacyPlanParams): Promise<QueryPlan> {
    const { projectDescription, memorySnapshot = '', sources, roundNumber = 1, maxRounds = 5 } = params
    if (!projectDescription || !projectDescription.trim()) {
      throw new QueryPlanError('projectDescription is required')
    }
    if (!Array.isArray(sources) || sources.length === 0) {
      throw new QueryPlanError('sources must be a non-empty array')
    }

    const promptFile = loadPrompt('agents/query_plan_legacy')
    const rendered = promptFile.render({
      project_description: projectDescription,
      memory_section: memorySnapshot || '（无历史记忆）',
      round_number: roundNumber,
      max_rounds: maxRounds,
      tool_reliability: '（客户端不传可靠性）',
      disabled_sources: '（由调用方过滤）',
      prev_stats_section: '',
    })

    let lastErr: unknown = null
    for (let attempt = 0; attempt < 2; attempt++) {
      try {
        const response = await this.llm.generate(rendered, {
          temperature: 0.2,
          response_format: { type: 'json_object' },
        })
        if (!response || !response.text) {
          lastErr = new Error('LLM returned empty')
          continue
        }
        const data = parseLegacyPlanJson(response.text)
        if (!data) {
          lastErr = new Error(`parse failed: ${response.text.slice(0, 150)}`)
          continue
        }
        const plan = this.buildQueryPlanFromLegacy(data, sources)
        if (plan) return plan
        lastErr = new Error('plan missing base_query')
      } catch (e) {
        lastErr = e
      }
    }
    throw new QueryPlanError(
      `legacyPlan failed after 2 attempts: ${(lastErr as Error)?.message ?? 'unknown'}`,
      lastErr,
    )
  }

  // ═══════════════════════════════════════════════════════════════════
  // Helpers
  // ═══════════════════════════════════════════════════════════════════

  private async fallbackToLegacy(
    params: AgenticPlanParams,
    reason: string,
  ): Promise<QueryPlan> {
    try {
      const plan = await this.legacyPlan({
        projectDescription: params.projectDescription,
        memorySnapshot: params.memorySnapshot,
        sources: params.sources,
      })
      return {
        ...plan,
        reasoning: `[fallback from agentic: ${reason}] ${plan.reasoning}`,
      }
    } catch (e) {
      // 三层兜底 3：legacy 也挂 → throw
      throw new QueryPlanError(
        `agentic + legacy both failed. agentic reason: ${reason}; legacy: ${(e as Error).message}`,
        e,
      )
    }
  }

  /** 把对话历史拼成单 string prompt（给不支持 native messages 的 provider）。 */
  private renderConversation(
    history: ReadonlyArray<{ role: string; content: string }>,
    systemPrompt: string,
  ): string {
    const parts = [systemPrompt, '']
    for (const m of history) {
      parts.push(`=== ${m.role.toUpperCase()} ===`)
      parts.push(m.content)
      parts.push('')
    }
    parts.push('=== ASSISTANT ===')
    return parts.join('\n')
  }

  /** 把 agent finalize 的 dict 转成 QueryPlan。返回 null 表示数据不可用。 */
  private buildQueryPlan(
    planData: Extract<AgentAction, { action: 'finalize' }>['plan'],
    sources: string[],
    mode: 'agentic' | 'legacy',
    iterations: number,
  ): QueryPlan | null {
    const baseQuery = (planData.base_query ?? '').trim()
    const clarificationNeeded = Boolean(planData.clarification_needed)
    const clarificationMessage = (planData.clarification_message ?? '').slice(0, 500)

    if (!baseQuery && !clarificationNeeded) return null

    // 年份消毒
    const currentYear = new Date().getFullYear()
    let yearTo = planData.year_to ?? null
    if (typeof yearTo !== 'number' || yearTo > currentYear + 1) yearTo = currentYear
    let yearFrom = planData.year_from ?? null
    if (yearFrom !== null && (typeof yearFrom !== 'number' || yearFrom < 1900)) {
      yearFrom = null
    }

    // language_scope 消毒
    let languageScope = (planData.language_scope ?? 'international') as
      | 'chinese_first'
      | 'international'
      | 'global'
    if (!['chinese_first', 'international', 'global'].includes(languageScope)) {
      languageScope = 'international'
    }

    const excludeTerms = Array.isArray(planData.exclude_terms) ? planData.exclude_terms : []

    // base_query 切词作为 keywords
    const keywords = (baseQuery || 'placeholder')
      .split(/\s+/)
      .filter((w) => w.length >= 2)
    if (keywords.length === 0) keywords.push(baseQuery || 'placeholder')

    const filters: Record<string, unknown> = {
      year_from: yearFrom,
      year_to: yearTo,
      language_scope: languageScope,
      exclude_terms: excludeTerms,
    }

    const perSource: Record<string, SourcePlan> = {}
    for (const s of sources) {
      perSource[s] = { keywords: [...keywords], filters: { ...filters } }
    }

    const rationale = (planData.rationale ?? '').slice(0, 300)
    const reasoning = clarificationNeeded
      ? `[clarification_needed] ${clarificationMessage} | ${rationale}`
      : rationale

    return {
      perSource,
      reasoning,
      iterations,
      meta: {
        baseQuery,
        chineseQuery: planData.chinese_query ?? null,
        yearFrom,
        yearTo,
        languageScope,
        excludeTerms,
        clarificationNeeded,
        clarificationMessage,
        mode,
      },
    }
  }

  /** 用 legacy LLM 输出（已验过 base_query）构造 QueryPlan。 */
  private buildQueryPlanFromLegacy(
    data: LegacyPlanJson,
    callerSources: string[],
  ): QueryPlan | null {
    return this.buildQueryPlan(
      {
        base_query: data.base_query,
        chinese_query: data.chinese_query,
        year_from: data.year_from,
        year_to: data.year_to,
        language_scope: data.language_scope,
        rationale: data.rationale,
        exclude_terms: data.exclude_terms,
        clarification_needed: false,
        clarification_message: '',
      },
      callerSources,
      'legacy',
      0,
    )
  }
}
